_source/_posts/2022-06-01-oauth-public-client-identity.md

```diff
@@ -0,0 +1,111 @@
+
+In order to do this, the API needs to know which client an access token was issued to. Typically the authorization server will embed the application's client_id in one of the claims of the access token so that it's accessible to the API when it validates the token. Of course this information is only as good as the authorization server's confidence that it was really that app making the token request. Again, if the client can authenticate itself, then this is reasonably reliable. But for public clients, there is much less certainty.
+
+Keep privileged access in mind when you're building APIs that limit features based on the application.
+
+
+## Frequently asked questions
+
+With these concerns in mind, let's answer a few quick questions about OAuth client impersonation.
+
+### Is client impersonation a new problem?
+
+Not at all! In fact, client impersonation [was mentioned in the original OAuth 2.0 spec](https://datatracker.ietf.org/doc/html/rfc6749#section-10.2) RFC 6749 in 2012! The next year, the "Threat Model and Security Considerations" document was finished and became RFC 6819. This document has [a large section]((https://datatracker.ietf.org/doc/html/rfc6749#section-10.2)) dedicated to the topic of client impersonation. That section is worth a read for more information on various ways to mitigate this risk.
+
+### Is client impersonation a problem unique to OAuth?
+
+Client impersonation is relevant to OAuth for the reasons previously mentioned, but those issues may also be relevant to systems that don't use OAuth at all. 
+
+This is a pervasive problem with mobile app games. Let's take a simple example of building a leaderboard service for a mobile game application. For the sake of the example, let's assume the app doesn't have any concept of user accounts, and wants to show a high score leaderboard in the game from all users of the app worldwide. In order for the game server to collect the high score data, it provides an API endpoint that the game can report scores to. The API would like to only accept POST requests from real instances of the game app, and not people who are trying to cheat and send fake scores to the API. 
+
+Without the support of the mobile operating system, there's nothing in the API call that can authenticate the mobile app, since any HTTP request the app makes would be possible for anyone to mimic outside of the application.
+
+### Does PKCE solve client impersonation?
+
+Nope! This is a common point of confusion, likely because of how PKCE was originally developed and rolled out. PKCE was originally described as a way for a mobile app to do the authorization code flow safely. But keep in mind this was back in 2015 when the previous advice was for mobile apps to use the Implicit flow, or the authorization code flow without credentials.
```

Comment:
```suggestion
Nope! This is a common point of confusion, likely because of how PKCE was originally developed and rolled out. PKCE was originally described as a way for a mobile app to do the authorization code flow safely. But keep in mind this was back in 2015 when mobile apps were still being advised to use the Implicit flow or the authorization code flow without credentials.
```
more clarifying/simplifying